### PR TITLE
Move signing functionality into Signing Service

### PR DESCRIPTION
**Pull request by janakitti**

### CODE REVIEW

File: src/app/core/signing.service.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
+import { Injectable } from '@angular/core';
+import { Fonts } from '../shared/interfaces';
+import { ActivatedRoute, Router } from '@angular/router';
+import { AuthService } from './auth.service';
```
Comment (ricetech) [inline src/app/core/signing.service.ts:4]: Remove unused imports
```suggestion
```

File: src/app/core/signing.service.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+  constructor() {
+    this.fontDict = FontService.fonts;
+    this.maxCharCount = MsgIoService.maxMessageLength;
+  }
+
+  /**
+   * Sets all class variables to their default values
+   */
+  resetDefaultValues() {
```
Comment (ricetech) [inline src/app/core/signing.service.ts:34]: Why is this a separate function?

Comment (ricetech) [inline src/app/core/signing.service.ts:34]: Why is this a separate method?

File: src/app/core/signing.service.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  public textColor;
+  public canSend;
+  public charCount;
+  public maxCharCount: number;
+  public charCountColor;
+
+  // font service stuff
+  public currentFont; // Used to select a font from fontDict
+  public currentFontFamily; // Used to CSS select the font
+  public currentFontDisplay; // Shown in the Font Dropdown menu
```
Comment (ricetech) [inline src/app/core/signing.service.ts:23]: Use strong typing
```suggestion
  public signingBoxText: string;
  public backgroundColor: string;
  public textColor: string;
  public canSend: boolean;
  public charCount: number;
  public maxCharCount: number;
  public charCountColor: string;

  // font service stuff
  public currentFont: string; // Used to select a font from fontDict
  public currentFontFamily: string; // Used to CSS select the font
  public currentFontDisplay: string; // Shown in the Font Dropdown menu
```

File: src/app/core/signing.service.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+   * @param font - Font selected from dropdown menu
+   */
+  selectFont(font: string) {
+    this.currentFont = font;
+    this.currentFontFamily = font + ',' + this.fontDict[font].backupFont;
+    this.currentFontDisplay = this.fontDict[font].fontFamily;
+  }
+
+  /**
+   * The following functions add text formatting characters around selected text within textbox.
```
Comment (ricetech) [inline src/app/core/signing.service.ts:56]: JSDoc is too wordy
```suggestion
   * Adds markdown formatting characters around selected text within textbox.
```

File: src/app/pages/signing/signing.component.ts [hunk 5]
```diff
@@ -16,29 +15,13 @@ import { Fonts } from '../../shared/interfaces';
...
-  public currentFont = 'Arial'; // Used to select a font from fontDict
-  public currentFontFamily = 'Arial, sans-serif'; // Used to CSS select the font
-  public currentFontDisplay = 'Arial'; // Shown in the Font Dropdown menu
-  public fontDict: Fonts;
-
-  constructor(private route: ActivatedRoute, private authService: AuthService, private msgIo: MsgIoService,
-              private router: Router) {
-    this.fontDict = FontService.fonts;
-  }
+  constructor(private route: ActivatedRoute, private authService: AuthService, public signService: SigningService,
```
Comment (ricetech) [inline src/app/pages/signing/signing.component.ts:23]: Why not `signingService`? `signService` sounds a bit off

File: src/app/core/signing.service.ts [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+  resetDefaultValues() {
+    this.signingBoxText = '';
+    this.backgroundColor = '#ffffff';
+    this.textColor = '#000000';
+    this.canSend = false;
+    this.charCount = 0;
+    this.charCountColor = '#bbbbbb';
+    this.currentFont = 'Arial'; // Used to select a font from fontDict
+    this.currentFontFamily = 'Arial, sans-serif'; // Used to CSS select the font
+    this.currentFontDisplay = 'Arial'; // Shown in the Font Dropdown menu
```
Comment (GuyInFridge) [inline src/app/core/signing.service.ts:43]:
```suggestion
    this.currentFont = 'Arial'; 
    this.currentFontFamily = 'Arial, sans-serif'; 
    this.currentFontDisplay = 'Arial'; 
```
Remove redundant comments (already mentioned in definition of these variables)

File: src/app/core/signing.service.ts [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+export class SigningService {
+  public signingBoxText;
+  public backgroundColor;
+  public textColor;
+  public canSend;
+  public charCount;
+  public maxCharCount: number;
+  public charCountColor;
+
+  // font service stuff
```
Comment (GuyInFridge) [inline src/app/core/signing.service.ts:20]:
```suggestion
  // Font service variables
```
Let's be a bit more professional, shall we? ;P

File: src/app/pages/signing/signing.component.html [hunk 8]
```diff
@@ -89,8 +93,8 @@
       </div>
       <div class="row">
         <markdown emoji
-                  [ngStyle]="{'background-color': backgroundColor, 'color': textColor, 'font-family': currentFontFamily}"
-                  id="preview-text" data="{{ signingBoxText }}"></markdown>
+                  [ngStyle]="{'background-color': signService.backgroundColor, 'color': signService.textColor, 'font-family': signService.currentFontFamily}"
+                  id="preview-text" data="{{ signService.signingBoxText }}"></markdown>
```
Comment (GuyInFridge) [inline src/app/pages/signing/signing.component.html:97]:
```suggestion
                  [ngStyle]="{'background-color': signService.backgroundColor, 'color': signService.textColor,
                  'font-family': signService.currentFontFamily}" id="preview-text"
                  data="{{ signService.signingBoxText }}"></markdown>
```
Over line length.

File: src/app/pages/signing/signing.component.ts (hunk 5 above)
Comment (GuyInFridge) [reply src/app/pages/signing/signing.component.ts:23]: I agree, that seems weird.

File: src/app/core/signing.service.ts (hunk 7 above)
Comment (janakitti) [reply src/app/core/signing.service.ts:18]: that wasn't me

File: src/app/core/signing.service.ts (hunk 2 above)
Comment (janakitti) [reply src/app/core/signing.service.ts:32]: We want any component to be able to reset the values of all the message properties.
If someone plays with the values on Home, we don't want that to show up on Signing when they go to sign, since the service is a singleton

File: src/app/pages/signing/signing.component.ts (hunk 5 above)
Comment (janakitti) [reply src/app/pages/signing/signing.component.ts:23]: Ok i'll change it, it just got a bit long in the html

File: src/app/core/signing.service.ts [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+  public canSend;
+  public charCount;
+  public maxCharCount: number;
+  public charCountColor;
+
+  // Font service variables
+  public currentFont; // Used to select a font from fontDict
+  public currentFontFamily; // Used to CSS select the font
+  public currentFontDisplay; // Shown in the Font Dropdown menu
+  public fontDict: Fonts;
```
Comment (ricetech) [inline src/app/core/signing.service.ts:22]: Use strong typing
```suggestion
  public signingBoxText: string;
  public backgroundColor: string;
  public textColor: string;
  public canSend: boolean;
  public charCount: number;
  public maxCharCount: number;
  public charCountColor: string;

  // font service stuff
  public currentFont: string; // Used to select a font from fontDict
  public currentFontFamily: string; // Used to CSS select the font
  public currentFontDisplay: string; // Shown in the Font Dropdown menu
```

File: src/app/pages/signing/signing.component.ts [hunk 10]
```diff
@@ -16,29 +15,13 @@ import { Fonts } from '../../shared/interfaces';
...
-  /** Default values */
-  public signingBoxText = '';
-  public backgroundColor = '#ffffff';
-  public textColor = '#000000';
-  public canSend = false;
-  public charCount = 0;
-  public maxCharCount: number;
-  public charCountColor = '#bbbbbb';
-  public name = '[friend name]';
+  public name = 'your friend\'s';
```
Comment (TommyHuang-dev) [inline src/app/pages/signing/signing.component.ts:18]:
```suggestion
  public name = 'your friend;
```
Remove the 's since we use the sentence "Type your message to {{ name }} below."

File: src/app/core/signing.guard.ts [hunk 11]
```diff
@@ -43,21 +43,24 @@ export class SigningGuard implements CanActivate {
       SigningGuard.key = linkStr.match(/key=([^&]*)/)[1];
     } catch (e) {
       // noinspection JSIgnoredPromiseFromCall
-      this.router.navigate([ '/invalid-link' ]);
+      this.router.navigate([ '/invalid-link' ]).catch(e => alert('Something went wrong, see the error below: \n' + e +
+        '\n please contact us at virtrolio.team@gmail.com if this problem persists.'));
```
Comment (ricetech) [inline src/app/core/signing.guard.ts:47]: Replace with standardized Error message
```suggestion
      this.router.navigate([ '/invalid-link' ]).catch(e => AuthService.displayError(e));
```

File: src/app/core/signing.guard.ts [hunk 12]
```diff
@@ -43,21 +43,24 @@ export class SigningGuard implements CanActivate {
...
     }
 
     /** Redirection based on authService.checkKey() & authService.isLoggedIn() */
     return this.authService.checkKey(SigningGuard.uid, SigningGuard.key).then(validKey => {
       if (validKey === false) {
         // noinspection JSIgnoredPromiseFromCall
-        this.router.navigate([ '/invalid-link' ]);
+        this.router.navigate([ '/invalid-link' ]).catch(e => alert('Something went wrong, see the error below: \n' + e +
+          '\n please contact us at virtrolio.team@gmail.com if this problem persists.'));
```
Comment (ricetech) [inline src/app/core/signing.guard.ts:55]: Replace with standardized Error message
```suggestion
      this.router.navigate([ '/invalid-link' ]).catch(e => AuthService.displayError(e));
```

File: src/app/core/signing.guard.ts [hunk 13]
```diff
@@ -43,21 +43,24 @@ export class SigningGuard implements CanActivate {
...
         return false;
       }
 
       this.msgIOService.checkForMessage(SigningGuard.uid).then((signed) => {
         if (signed) {
           // noinspection JSIgnoredPromiseFromCall
-          this.router.navigate([ '/rejecc' ]);
+          this.router.navigate([ '/rejecc' ]).catch(e => alert('Something went wrong, see the error below: \n' + e +
+            '\n please contact us at virtrolio.team@gmail.com if this problem persists.'));
```
Comment (ricetech) [inline src/app/core/signing.guard.ts:63]: Replace with standardized Error message
```suggestion
          this.router.navigate([ 'rejecc' ]).catch(e => AuthService.displayError(e));
```

File: src/app/core/signing.guard.ts [hunk 14]
```diff
@@ -69,13 +72,16 @@ export class SigningGuard implements CanActivate {
         return true;
       } else {
         // noinspection JSIgnoredPromiseFromCall
-        this.router.navigate([ '/signing-auth-redirect' ], { queryParams: { uid: SigningGuard.uid, key: SigningGuard.key } });
+        this.router.navigate([ '/signing-auth-redirect' ], { queryParams: { uid: SigningGuard.uid, key: SigningGuard.key }} )
+          .catch(e => alert('Something went wrong, see the error below: \n' + e + '\n please contact us at virtrolio.team@gmail.com if this ' +
+            'problem persists.'));
```
Comment (ricetech) [inline src/app/core/signing.guard.ts:77]: Replace with standardized Error message
```suggestion
        this.router.navigate([ '/signing-auth-redirect' ], { queryParams: { uid: SigningGuard.uid, key: SigningGuard.key }} )
          .catch(e => AuthService.displayError(e));
```

File: src/app/core/signing.guard.ts [hunk 15]
```diff
@@ -69,13 +72,16 @@ export class SigningGuard implements CanActivate {
...
         return false;
       }
     })
       .catch(() => {
         // noinspection JSIgnoredPromiseFromCall
-        this.router.navigate([ '/invalid-link' ]);
+        this.router.navigate([ '/invalid-link' ]).catch(e => alert('Something went wrong, see the error below: \n' + e +
+          '\n please contact us at virtrolio.team@gmail.com if this problem persists.'));
```
Comment (ricetech) [inline src/app/core/signing.guard.ts:84]: Replace with standardized Error message
```suggestion
        this.router.navigate([ '/invalid-link' ]).catch(e => AuthService.displayError(e));
```

File: src/app/pages/your-virtrolio/your-virtrolio.component.html [hunk 16]
```diff
@@ -90,9 +91,17 @@ <h5 class="modal-title" id="send-message-modal-title">You're barking up the wron
       </div>
 
       <!-- Modal Body -->
-      <div class="modal-body" id='send-message-body'>
-        To sign a friend's virtrolio, they need to give you their 'sharing link'! To access <em>your</em>
-        sharing link, click the 'Sharing Link' button on this page!
+      <div class="modal-body sharing-link-body">
+        <p>
+          To sign a friend's virtrolio, they need to give you their 'sharing link'! You can plug in your friend's sharing
```
Comment (ricetech) [inline src/app/pages/your-virtrolio/your-virtrolio.component.html:96]: Unclear wording
```suggestion
          To sign a friend's virtrolio, they need to give you their 'sharing link'! You can paste your friend's sharing
```

File: src/app/pages/your-virtrolio/your-virtrolio.component.html [hunk 17]
```diff
@@ -90,9 +91,17 @@ <h5 class="modal-title" id="send-message-modal-title">You're barking up the wron
...
+          link below or directly into your browser search bar. To access <em>your</em> sharing link, click the 'Sharing Link' button on this page!
+        </p>
+        <label id="input-sharing-label" for="input-sharing-field" style="display:none">Visit friend link</label>
+        <input type="text" class="align-middle" #inputSharing value="Input your friend's link here"
+               (click)="copyLink(inputSharing)" id="input-sharing-field" [(ngModel)]='visitLink'>
+        <button type="button" id="button-copy" class="btn btn-primary align-middle" data-dismiss="modal" aria-label="Close"
+                (click)="navigateToLink()">Let's go
```
Comment (ricetech) [inline src/app/pages/your-virtrolio/your-virtrolio.component.html:103]: Unnecessary wordiness
```suggestion
                (click)="navigateToLink()">Go
```

File: src/app/pages/your-virtrolio/your-virtrolio.component.ts [hunk 18]
```diff
@@ -44,6 +48,23 @@ export class YourVirtrolioComponent implements OnInit {
...
+  navigateToLink() {
+    if (this.visitLink) {
+      try {
+        // noinspection JSIgnoredPromiseFromCall
+        this.visitLinkUID = this.visitLink.match(/uid=([^&]*)/)[1];
+        this.visitLinkKEY = this.visitLink.match(/key=([^&]*)/)[1];
+        window.location.href = '/signing?uid=' + this.visitLinkUID + '&key=' + this.visitLinkKEY;
+      } catch (e) {
+        this.router.navigate(['/invalid-link']).catch(e => alert('Something went wrong, see the error below: \n' + e +
+            '\n please contact us at virtrolio.team@gmail.com if this problem persists.'));
```
Comment (ricetech) [inline src/app/pages/your-virtrolio/your-virtrolio.component.ts:63]:
```suggestion
        this.router.navigate(['/invalid-link']).catch(e => AuthService.displayError(e));
```

File: src/app/app-routing.module.ts [hunk 19]
```diff
@@ -55,7 +56,8 @@ const routes: Routes = [
     path: 'signing',
     component: SigningComponent,
     canActivate: [ SigningGuard ],
-    resolve: { user: LoginResolver }
+    resolve: { user: LoginResolver },
+    canDeactivate: [PendingChangesGuard]
```
Comment (ricetech) [inline src/app/app-routing.module.ts:60]: Spacing
```suggestion
    canDeactivate: [ PendingChangesGuard ]
```

File: src/app/core/pending-changes.guard.ts [hunk 20]
```diff
@@ -0,0 +1,22 @@
+import { CanDeactivate } from '@angular/router';
+import { Injectable } from '@angular/core';
+import { Observable } from 'rxjs';
+
+export interface ComponentCanDeactivate {
+  canDeactivate: () => boolean | Observable<boolean>;
+}
+
+// SHOUTOUT TO https://stackoverflow.com/a/41187919/13737272
```
Comment (ricetech) [inline src/app/core/pending-changes.guard.ts:9]:
```suggestion
```

File: src/app/core/pending-changes.guard.ts [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+
+@Injectable()
+export class PendingChangesGuard implements CanDeactivate<ComponentCanDeactivate> {
+  canDeactivate(component: ComponentCanDeactivate): boolean | Observable<boolean> {
+    // if there are no pending changes, just allow deactivation; else confirm first
+    return component.canDeactivate() ?
+      true :
+      // NOTE: this warning message will only be shown when navigating elsewhere within your angular app;
+      // when navigating away from your angular app, the browser will show a generic warning message
+      // see http://stackoverflow.com/a/42207299/7307355
```
Comment (ricetech) [inline src/app/core/pending-changes.guard.ts:19]:
```suggestion
```

File: src/app/core/pending-changes.guard.ts [hunk 22]
```diff
@@ -0,0 +1,22 @@
...
+}
+
+// SHOUTOUT TO https://stackoverflow.com/a/41187919/13737272
+
+@Injectable()
+export class PendingChangesGuard implements CanDeactivate<ComponentCanDeactivate> {
+  canDeactivate(component: ComponentCanDeactivate): boolean | Observable<boolean> {
+    // if there are no pending changes, just allow deactivation; else confirm first
+    return component.canDeactivate() ?
+      true :
```
Comment (ricetech) [inline src/app/core/pending-changes.guard.ts:16]:
```suggestion
      true:
```

File: src/app/core/signing.guard.ts [hunk 23]
```diff
@@ -43,21 +43,21 @@ export class SigningGuard implements CanActivate {
...
-        this.router.navigate([ '/invalid-link' ]);
+      this.router.navigate([ '/invalid-link' ]).catch(e => AuthService.displayError(e));
         return false;
       }
 
       this.msgIOService.checkForMessage(SigningGuard.uid).then((signed) => {
         if (signed) {
           // noinspection JSIgnoredPromiseFromCall
-          this.router.navigate([ '/rejecc' ]);
+          this.router.navigate([ 'rejecc' ]).catch(e => AuthService.displayError(e));
```
Comment (ricetech) [inline src/app/core/signing.guard.ts:60]: Missing slash
```suggestion
          this.router.navigate([ '/rejecc' ]).catch(e => AuthService.displayError(e));
```

File: src/app/core/pending-changes.guard.ts [hunk 24]
```diff
@@ -0,0 +1,20 @@
...
+
+@Injectable()
+export class PendingChangesGuard implements CanDeactivate<ComponentCanDeactivate> {
+  canDeactivate(component: ComponentCanDeactivate): boolean | Observable<boolean> {
+    // if there are no pending changes, just allow deactivation; else confirm first
+    return component.canDeactivate() ?
+      true:
+      // NOTE: this warning message will only be shown when navigating elsewhere within your angular app;
+      // when navigating away from your angular app, the browser will show a generic warning message
+      confirm('WARNING: You have unsaved changes. Press Cancel to go back and save these changes, or OK to lose these changes.');
```
Comment (ricetech) [inline src/app/core/pending-changes.guard.ts:18]: Wording doesn't match our service
```suggestion
      confirm('WARNING: Your message has not been sent yet. Press Cancel to go back to complete and send your message, or OK to discard your message.');
```

File: src/app/core/pending-changes.guard.ts [hunk 25]
```diff
@@ -0,0 +1,20 @@
...
+import { Injectable } from '@angular/core';
+import { Observable } from 'rxjs';
+
+export interface ComponentCanDeactivate {
+  canDeactivate: () => boolean | Observable<boolean>;
+}
+
+
+@Injectable()
+export class PendingChangesGuard implements CanDeactivate<ComponentCanDeactivate> {
```
Comment (GuyInFridge) [inline src/app/core/pending-changes.guard.ts:11]: @TommyHuang-dev could you add proper TS Documentation to this canDeactivate function instead of having the comments inside the function?

File: src/app/pages/signing/signing.component.ts [hunk 26]
```diff
@@ -48,65 +33,34 @@ export class SigningComponent implements OnInit {
...
-   * Update character count value and text colour
-   * @param textbox - textbox in which user types.
-   */
-  updateCount(textbox: HTMLTextAreaElement) {
-    this.charCount = textbox.value.length;
-    this.charCountColor = (this.charCount > this.maxCharCount) ? '#EE1111' : '#b0b0b0';
-    this.canSend = (0 < this.charCount && this.charCount <= this.maxCharCount);
+  // @HostListener allows us to guard against browser refresh, close, etc.
+  @HostListener('window:beforeunload')
+  canDeactivate(): Observable<boolean> | boolean {
```
Comment (GuyInFridge) [inline src/app/pages/signing/signing.component.ts:42]: @TommyHuang-dev can you change this to proper TS documentation above the function instead of within?

File: src/app/pages/your-virtrolio/your-virtrolio.component.ts [hunk 27]
```diff
@@ -44,6 +48,22 @@ export class YourVirtrolioComponent implements OnInit {
     this.linkReady = true;
   }
 
+  /**
+   * Attempt to navigate by router to link inputted in 'send a message' input field
```
Comment (GuyInFridge) [inline src/app/pages/your-virtrolio/your-virtrolio.component.ts:52]:
```suggestion
   * Attempts to navigate by router to a sharing link based on what is inputted in the 'send a message' input field (query params extracted with regEx if possible)
```
